core/src/proof.rs

```diff
@@ -0,0 +1,135 @@
+//! Proving and verifying inclusion, non-inclusion, and updates to the trie.
+
+use crate::trie::{InternalData, KeyPath, LeafData, Node, NodeHasher, NodeHasherExt, TERMINATOR};
+
+use alloc::vec::Vec;
+use bitvec::prelude::*;
+
+/// Generic cursor over binary trie storage.
+pub trait Cursor {
+    /// The current position of the cursor, expressed as a bit-path and length. Bits after the
+    /// length are irrelevant.
+    fn position(&self) -> (KeyPath, u8);
+    /// The current node.
+    fn node(&self) -> Node;
+
+    /// Peek at the children of this node, if it is an internal node.
+    fn peek_children(&self) -> Option<(&Node, &Node)>;
+
+    /// Traverse to the left child of this node, if it is an internal node. No-op otherwise.
+    fn traverse_left_child(&mut self);
+    /// Traverse to the right child of this node, if it is an internal node. No-op otherwise.
+    fn traverse_right_child(&mut self);
+    /// Traverse upwards by d bits. No-op if d is greater than the current position length.
+    fn traverse_parents(&mut self, d: u8);
+}
+
+/// Sibling nodes recorded while looking up some path, in ascending order by depth.
+#[derive(Clone)]
+pub struct Siblings(pub Vec<Node>);
+
+/// A proof of some particular path through the trie.
+#[derive(Clone)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct PathProof {
+    /// The terminal node encountered when looking up a key. This is always either a terminator or
+    /// leaf.
+    pub terminal: Option<LeafData>,
+    /// Sibling nodes encountered during lookup, in ascending order by depth.
+    pub siblings: Siblings,
+}
+
+impl PathProof {
+    /// Verify this path proof.
+    ///
+    /// Provide the root node and a key path. The key path can be any key that results in the
+    /// lookup of the terminal node.
+    pub fn verify<H: NodeHasher>(
+        &self,
+        key_path: &KeyPath,
+        root: Node,
+    ) -> Result<VerifiedPathProof, ()> {
+        // TODO, proper error
+        let mut cur_node = match &self.terminal {
+            None => TERMINATOR,
+            Some(leaf_data) => H::hash_leaf(&leaf_data),
+        };
+
+        let relevant_path = &key_path.view_bits::<Msb0>()[..self.siblings.0.len()];
+        for (bit, &sibling) in relevant_path.iter().by_vals().rev().zip(&self.siblings.0) {
+            let (left, right) = if bit {
+                (sibling, cur_node)
+            } else {
+                (cur_node, sibling)
+            };
+
+            let next = InternalData {
+                left: left.clone(),
+                right: right.clone(),
+            };
+            cur_node = H::hash_internal(&next);
```

Comment:
Or alternatively, `hash_internal` just takes `left` and `right`.

Reply:
may be better, will punt on that here.